src/uri-template.ts

```diff
@@ -1,24 +1,57 @@
-class UriTemplate {
+import * as path from 'path';
+
+export interface PythonScript {
+    content: string;
+    checksum: number;
+    fullPath: string;
+}
+
+export class UriTemplate {
     private readonly _templateString: string = '';
-    public constructor(fileName: string, pythonScriptPath: string, fileExtensions: string) {
+    /**
+     * @param dataPluginName name of DataPlugin and uri
+     * @param pythonScript full file path to script or object of file content, crc32 and name
+     * @param fileExtensions supported file extensions of the DataPlugin
+     */
+    public constructor(
+        dataPluginName: string,
+        pythonScript: string | PythonScript,
+        fileExtensions: string
+    ) {
         this._templateString =
-            `<usireginfo><storetype name="${fileName}">` +
+            `<usireginfo><storetype name="${dataPluginName}">` +
             '<type>python</type>' +
-            `<alias>${fileName}</alias>` +
-            `<description>${fileName}</description>` +
+            `<alias>${dataPluginName}</alias>` +
+            `<description>${dataPluginName}</description>` +
             '<filepath>uspTdmMarshaller.dll</filepath>' +
             '<exportsupported>NO</exportsupported>' +
-            '<caching>YES</caching>' +
-            `<easypluginparam><![CDATA[<dllpath>@USIBINDIR@\\PythonMarshaller\\uspPythonMarshaller.dll</dllpath><script>${pythonScriptPath}</script>]]></easypluginparam>` +
-            '<querysupported>0</querysupported>' +
-            '<fastloadsupported>0</fastloadsupported>' +
-            `<filefilters extension="${fileExtensions}"><description>${fileName} Files (${fileExtensions})</description></filefilters>` +
-            '<platform>x64</platform></storetype></usireginfo>';
+            '<caching>YES</caching>';
+
+        if (typeof pythonScript === 'string') {
+            this._templateString +=
+                `<easypluginparam><![CDATA[<dllpath>@USIBINDIR@\\PythonMarshaller\\uspPythonMarshaller.dll</dllpath><script>${pythonScript}</script>]]></easypluginparam>` +
+                '<querysupported>0</querysupported>' +
+                '<fastloadsupported>0</fastloadsupported>' +
+                `<filefilters extension="${fileExtensions}"><description>${dataPluginName} Files (${fileExtensions})</description></filefilters>` +
+                '<platform>x64</platform></storetype></usireginfo>';
+        } else {
+            const pyScriptName: string = path.basename(pythonScript.fullPath);
+            this._templateString +=
```

Comment:
Actually this looks like some parts of the template XML are duplicated in the if-clause and else-clause. If there isn't a better way of implemented this, I would feel better if the test coverage of this code is pretty high.

Reply:
thanks, this makes sense. I guess I want to avoid xml builders for now but I definitely had to restructure the code so it's more clear what is the difference in the if- and else-clause. I also added two tests that ensure the output is as expected.

Reply:
Thanks, looking much better now.